presto-main/src/main/java/com/facebook/presto/operator/PagesIndex.java

```diff
@@ -213,13 +214,19 @@ public void addPage(Page page)
         positionCount += page.getPositionCount();
 
         int pageIndex = (channels.length > 0) ? channels[0].size() : 0;
+        ReferenceCountMap referenceCountMap = new ReferenceCountMap();
         for (int i = 0; i < channels.length; i++) {
             Block block = page.getBlock(i);
             if (eagerCompact) {
                 block = block.copyRegion(0, block.getPositionCount());
             }
             channels[i].add(block);
-            pagesMemorySize += block.getRetainedSizeInBytes();
+
+            block.retainedBytesForEachPart((object, size) -> {
```

Comment:
My memory is that this is unbelievably expensive, which is why we only did this very sparingly.  Did you do some performance analysis?

Reply:
Thanks for clarification. I was wondering if there are some issues with using `retainedBytesForEachPart` method.

> Did you do some performance analysis?

I will add some benchmarks of building/compacting `PagexIndex` with pages containing high number of channels.

> Also, if we have to use this interface, run benchmarks with huge inputs (e.g., 1 billion distinct blocks or something like that) to show the effect.

This `ReferenceCountMap` exists only for the duration of the page (it should be the same in `compact()` too), so the overhead factor should be linear with the number of page channels. I will add such benchmarks.